Fix generate argument parsing

erichte-ibm opened the pull request:
Fixes #80 

Argp calls the `parse_options()` function for each argument, therefore the logic to reject the append flag on a PK was evaluated for other arguments. Move the append flag reject logic outside of the argument parser, since it only needs to be evaluated once anyway, after all arguments have been read.

---

While solving the above, I noticed that the positional format specifier argument parsing logic was fairly... lacking. It led to some confusing error messages, as the original test case was using the old `-a 0` syntax for specifying the append header. The `0` was treated as a positional argument, and therefore was expected to be in the `c:a` format. Since it was _not_, it overwrote the previously valid `c:a` format specifier with `NULL`s, leading to an error in the format specifier, NOT in an unexpected argument.

The second commit addresses this, by adding some immediate enforcement of the format specifier, returning an error if it does not fit, or if another positional argument is supplied.

nick-child-ibm commented on backends/guest/guest_svc_generate.c:
```diff
@@ -493,14 +506,19 @@ static int parse_options(int key, char *arg, struct argp_state *state)
 			break;
-		else if (args->input_form == NULL || args->output_form == NULL)
-			prlog(PR_ERR, "ERROR: incorrect '<input_format>:<output_format>', see "
-				      "usage...\n");
```
This final check is necessary otherwise the user can get away with not setting the `in:out` argument:
```
$ ./bin/secvarctl-dbg -m guest generate -i sbat_in -o sbat.auth 
AddressSanitizer:DEADLYSIGNAL
=================================================================
==566168==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000000 (pc 0x0000004377a9 bp 0x7ffe8abfeae0 sp 0x7ffe8abfeab0 T0)
==566168==The signal is caused by a READ memory access.
==566168==Hint: address points to the zero page.
    #0 0x4377a8 in parse_options backends/guest/guest_svc_generate.c:525
    #1 0x7fe995198892 in argp_parse (/lib64/libc.so.6+0x13a892)
    #2 0x438dbc in guest_generate_command backends/guest/guest_svc_generate.c:666
    #3 0x4069b4 in main /home/nchild/IBM/secvarctl-stuff/erichte-guest-devel/secvarctl/secvarctl.c:250
    #4 0x7fe995098d84 in __libc_start_main (/lib64/libc.so.6+0x3ad84)
    #5 0x4031dd in _start (/home/nchild/IBM/secvarctl-stuff/erichte-guest-devel/secvarctl/bin/secvarctl-dbg+0x4031dd)

AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV backends/guest/guest_svc_generate.c:525 in parse_options
==566168==ABORTING
```
This can avoided by adding back this conditional:
```
$ ./bin/secvarctl-dbg -m guest generate -i sbat_in -o sbat.auth 
ERROR: incorrect '<input_format>:<output_format>', see usage...
Usage: secvarctl -m guest generate [OPTION...]
            <input_format>:<output_format> -i <input_file> -o <output_file>
  or:  secvarctl -m guest generate [OPTION...]
            reset -i <input_file> -o <output_file>
Try `secvarctl -m guest generate --help' or `secvarctl -m guest generate
--usage' for more information.
failed during argument parsing
RESULT: FAILURE
```

erichte-ibm replied:
Yep, that's an oversight on my part. Re-added the check, and tweaked the message to include potentially missing the argument.

erichte-ibm commented:
Updated:
 - re-added the check
 - added test cases to hopefully catch these in the future
 - added a new commit that allows the test framework to more properly catch exits via `SIGSEGV`, `SIGABRT`, etc
